c++/src/kj/exception.c++

```diff
@@ -1070,9 +1076,30 @@ private:
 };
 
 ExceptionCallback& getExceptionCallback() {
-  static ExceptionCallback::RootExceptionCallback defaultCallback;
+  static auto defaultCallback = new ExceptionCallback::RootExceptionCallback();
+  // We allocate on the heap because some objects may throw in their destructors. If those objects
+  // had static storage, they might get fully constructed before the root callback. If they however
+  // then throw an exception during destruction, there would be a lifetime issue because their
+  // destructor would end up getting registered after the root callback's destructor. One solution
+  // is to just leak this pointer & allocate on first-use. The cost is that the initialization is
+  // mildly more expensive (+ we need to annotate sanitizers to ignore the problem). A great
+  // compiler annotation that would simply things would be one that allowed static variables to have
+  // their destruction omitted wholesale. That would allow us to avoid the heap but still have the
```

Comment:
> one that allowed static variables to have their destruction omitted wholesale.

I came up with this. Not sure about the smell, though -- I'm happy with the leak solution, too.

```cpp
ExceptionCallback& getExceptionCallback() {
  static union {
    ExceptionCallback::RootExceptionCallback defaultCallback;
  };

  // Use a second static to guarantee we call ctor() once.
  static auto initialized = [&defaultCallback]() { kj::ctor(defaultCallback); return true; }();

  ExceptionCallback* scoped = threadLocalCallback;
  return scoped != nullptr ? *scoped : defaultCallback;
}
```

Reply:
ooooo. I've never seen a trick like that. Filing that away. @kentonv do you like this better?

Reply:
I don't have a strong opinion.

The IIFE does seem like a nice way to accomplish one-time static initialization, compared to a separate global function.

Reply:
I thought about it & I know I tested the version as written but I don't have a convenient set up at the moment to retest the new code snippet. I'll leave a todo for now & implement this in a a follow-on PR